### PR TITLE
Add UI functionality to duplicate the user data

### PR DESCRIPTION
**Pull request by levitte**
This can be used by engines that need to retain the data for a longer time
than just the call where this user data is passed.

##### Checklist

- [x] documentation is added or updated

### CODE REVIEW

**Comment by levitte** [comment]
@dwmw2, I believe you're interested in this

**Comment by dwmw2** [comment]
Staring at it on my.phone probably doesn't help but... Where does the 'destroy' function originally come from? Shouldnt it be passed in from the app as an additional argument to the 'set duplicator' call. Dup and free functions are a pair. (And if it's recounting are just Inc and dec)

**Comment by levitte** [comment]
Very good point, I'll fix

**Comment by dwmw2** [comment]
I might even go so far as to suggest that the add_user_data function should take all three. To add user data without the dup/free funcs you should need to explicitly pass NULL

**Comment by levitte** [comment]
I went the other way and made the destructor a UI_METHOD function as well.

**Comment by dwmw2** [comment]
... and the UI data so added is always destroyed by OpenSSL for consistency.

**Comment by dwmw2** [comment]
LGTM

File: doc/man3/UI_create_method.pod [hunk 1]
```diff
@@ -5,9 +5,11 @@
...
 UI_create_method, UI_destroy_method, UI_method_set_opener,
 UI_method_set_writer, UI_method_set_flusher, UI_method_set_reader,
-UI_method_set_closer, UI_method_set_prompt_constructor,
-UI_method_set_ex_data, UI_method_get_opener, UI_method_get_writer,
-UI_method_get_flusher, UI_method_get_reader, UI_method_get_closer,
+UI_method_set_closer, UI_method_set_data_duplicator,
+UI_method_set_prompt_constructor, UI_method_set_ex_data,
+UI_method_get_opener, UI_method_get_writer, UI_method_get_flusher,
+UI_method_get_reader, UI_method_get_closer,
+UI_method_get_data_duplicator, UI_method_get_data_destructor,
```
Comment (mattcaswell) [inline doc/man3/UI_create_method.pod]: Update HISTORY section?

File: doc/man3/UI_new.pod [hunk 2]
```diff
@@ -7,9 +7,9 @@ UI_new, UI_new_method, UI_free, UI_add_input_string, UI_dup_input_string,
 UI_add_verify_string, UI_dup_verify_string, UI_add_input_boolean,
 UI_dup_input_boolean, UI_add_info_string, UI_dup_info_string,
 UI_add_error_string, UI_dup_error_string, UI_construct_prompt,
-UI_add_user_data, UI_get0_user_data, UI_get0_result, UI_process,
-UI_ctrl, UI_set_default_method, UI_get_default_method, UI_get_method,
-UI_set_method, UI_OpenSSL, UI_null - user interface
+UI_add_user_data, UI_dup_user_data, UI_get0_user_data, UI_get0_result,
```
Comment (mattcaswell) [inline doc/man3/UI_new.pod]: HISTORY?

File: crypto/ui/ui_lib.c [hunk 3]
```diff
@@ -387,10 +390,38 @@ char *UI_construct_prompt(UI *ui, const char *object_desc,
...
+    if ((ui->flags & UI_FLAG_DUPL_DATA) != 0) {
+        ui->meth->ui_destroy_data(ui, old_data);
+        old_data = NULL;
+    }
     ui->user_data = user_data;
+    ui->flags &= ~UI_FLAG_DUPL_DATA;
     return old_data;
 }
 
+int UI_dup_user_data(UI *ui, void *user_data)
```
Comment (mattcaswell) [inline crypto/ui/ui_lib.c]: Should this really be called UI_set1_user_data()?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/ui/ui_lib.c (hunk 3 above)
Comment (levitte) [inline crypto/ui/ui_lib.c]: That would be inconsistent with the rest of the UI API.  You might want to argue that the API should be "modernised", but that's for 1.2.0, not for now.

Background: the whole set0 / set1 semantics came after UI, as far as I recall, or was at least not settled and well known throughout the whole team at the time.

File: doc/man3/UI_create_method.pod (hunk 1 above)
Comment (levitte) [inline doc/man3/UI_create_method.pod]: Done

File: doc/man3/UI_new.pod (hunk 2 above)
Comment (levitte) [inline doc/man3/UI_new.pod]: Done

File: crypto/ui/ui_lib.c (hunk 3 above)
Comment (mattcaswell) [inline crypto/ui/ui_lib.c]: How about add1 then? There are quite a few instances of that in the public APIs already.

Comment (levitte) [inline crypto/ui/ui_lib.c]: How about looking at the rest of ui.h first? There is quite a number of add/dup pairs, why should this particular function differ?

Comment (mattcaswell) [inline crypto/ui/ui_lib.c]: Hmm. Ok.

**Review by mattcaswell — APPROVED** [review]

**Comment by levitte** [comment]
Merged. 545360c4dfbf90f63918be3d98e344adc7ca2d7f

Thanks!

**Pull request closed without merging**